### ISSUE TITLE
v3 development

### ISSUE THREAD

**Issue by rmg** [go-yaml/yaml#363]
@niemeyer I've seen you reference v3 development in several issues over the last few weeks but I can't find any sign of where to look for it. Where might one look to help out?

Sorry for what might end up a click-bait title. It seemed more appropriate than asking as a comment on one of the various issues that mentions v3.

**Comment by zbindenren** [comment]
Since #139 is going to be solved in v3, it would maybe be a good idea to create a v3 branch?

**Comment by cking** [comment]
am i right to assume that go-yaml is gone pretty stale?

last commit on v2 branch is from march (it is september basically rn) and the mentioned v3 is nowhere to be seen?

i really prefer yaml over its alternatives (mostly toml) but the missing comment support is really bothering me. for now i will switch over to toml, but i hope to be able to switch back at some point...

**Comment by niemeyer** [comment]
Please feel free to switch if it suits your needs. It's a nicer and simpler format in many senses, and although v3 is in the works, we won't have the ability to manipulate comments with it yet.

**Comment by cking** [comment]
i know about the last part.
also i prefer the yaml syntax over toml.
it is just that you cant see any progress, or anything. it feels a bit weird.
is there any kind of development repo you can watch?

**Comment by niemeyer** [comment]
No, I've been developing privately as I evolve an independent data manipulation package that I also hope to get to light soon. Hopefully they'll both be out in a short while.

**Comment by andreynering** [comment]
Two comments:

- Since go modules is now a thing, I think it makes sense to make v3 target `github.com/go-yaml/yaml/v3` instead of `gopkg.in/yaml.v3`;
- What is currently called "strict mode" should be the default, but maybe it should also allow a more loose mode (the current default).

**Comment by fgimian** [comment]
I'm extremely sorry to ask, but was just curious how progress on v3 is going?  I'm very keen to have the ability to define custom tag handlers and such similar to PyYAML which I believe was coming to v3.

Thanks for your hard work on this library
Fotis

**Comment by challarao** [comment]
Is there a feature list for v3 release version?

**Comment by niemeyer** [comment]
There will be a comprehensive list of all the changes so we'll be able to know what's interesting and what changed and needs to be ported. I'm trying to keep compatibility where that's possible (e.g. the Unmarshaler interface now takes a Node, but at least in v3 the old unmarshaler will still work so things are easier to port over).

**Comment by niemeyer** [comment]
> I'm extremely sorry to ask, but was just curious how progress on v3 is going? I'm very keen to have the ability to define custom tag handlers and such similar to PyYAML which I believe was coming to v3.

I've been running down a long list of small issues for quite a while. Things such as replacing the strict mode behavior with independent flags that allow breaking on known fields and/or duplicate keys, supporting custom indentation and fixing the bad cases that happened with that, supporting marshaling of anchors which is something we didn't have to worry about until now that we have an intermediate abstraction, making timestamps _always_ unmarshal into time.Time (we couldn't with v2 due to backward compatibility reasons), and so on.

It's coming, I promise. This is just taking a while because it's work I need to fit in the middle of everything else.

**Comment by thomasf** [comment]
@niemeyer  Any chance that we can have byte offset for start/end mark in `Node` as well as `Line` and `Column`? I'm looking for some library that can replace some simpler value types in some locations in ynl files while guaranteeing not to change anything but those values.  `event.start_mark.index` and   `event.end_mark.index`  would be perfect if I read the code correctly.

**Comment by laverya** [comment]
As far as I can tell @niemeyer v3 is looking good enough to make a release from - are there any known blockers there, besides making up the changelog? It would be nice to move over the default branch from v2 to v3, too.

I think it would be best to release once it's stable and make small improvement releases afterwards, but I don't know how you prefer to do things.

**Comment by laverya** [comment]
> What is currently called "strict mode" should be the default, but maybe it should also allow a more loose mode (the current default).

I will second the call for an `UnmarshalPermissive` or `UnmarshalUnsafe` function - otherwise this upgrade becomes a breaking change for quite a few apps that accept user input
